### PR TITLE
chore(xsnap): Upgrade xsnap-native

### PR DESCRIPTION
**Pull request by kriskowal**
Upgrades xsnap native to Moddable's finest.
https://github.com/agoric-labs/xsnap-pub/pull/20

### CODE REVIEW

File: packages/xsnap/test/test-xs-perf.js [hunk 1]
```diff
@@ -48,7 +48,7 @@ test('meter details', async t => {
 
   t.like(
     meters,
-    { compute: 1_380_192, allocate: 42_074_144 },
+    { compute: 1_380_187, allocate: 42_074_144 },
     'compute, allocate meters should be stable; update METER_TYPE?',
```
Comment (dckc) [inline packages/xsnap/test/test-xs-perf.js:52]: I don't see a change to `METER_TYPE`. Did I miss it?

**Comment by dckc** [comment]
Looking at the XS diffs (`git diff 55cc73f..d7895e4 xs/`)

### Minimum Mac OS version

Does this imply any updates to our docs about Mac OS requirements?

```
-MACOS_VERSION_MIN ?= -mmacosx-version-min=10.7
+MACOS_VERSION_MIN ?= -mmacosx-version-min=10.10
```

### Misc.

`fxAllocateChunks` is deleted from `xsHost.c`? ah... moved to `xs/platforms/mc/xsHosts.c`.

What's `mxRun`? In `xsBigInt.c`:

```C
+#ifdef mxRun
+               if ((t->size > 1) || (t->data[0] > 0xFFFF))
+                       mxRangeError("too big exponent");
+#endif
```

handy:

```
 #define XS_MAJOR_VERSION 11
-#define XS_MINOR_VERSION 5
-#define XS_PATCH_VERSION 0
+#define XS_MINOR_VERSION 7
+#define XS_PATCH_VERSION 1
```

### diff stats

`git diff --stat 55cc73f..d7895e4 xs/`

```
 xs/includes/xs.d.ts             |   59 +-
 xs/makefiles/lin/xst.mk         |   10 +-
 xs/makefiles/mac/xst.mk         |   24 +-
 xs/makefiles/win/xst.mak        |   12 +
 xs/platforms/esp/xsHost.c       |  715 +----
 xs/platforms/esp/xsHost.h       |   30 +-
 xs/platforms/esp/xsPlatform.c   |   14 +-
 xs/platforms/esp/xsPlatform.h   |    6 +
 xs/platforms/lin_xs.h           |    4 +
 xs/platforms/mac_xs.h           |    3 +
 xs/platforms/mc/xsHosts.c       |  636 ++++
 xs/platforms/mc/xsHosts.h       |   57 +
 xs/platforms/pico/xsHost.c      |  707 +++++
 xs/platforms/pico/xsHost.h      |  425 +++
 xs/platforms/pico/xsPlatform.c  |  722 +++++
 xs/platforms/pico/xsPlatform.h  |  169 +
 xs/platforms/win_xs.c           |   12 +-
 xs/platforms/win_xs.h           |    4 +
 xs/sources/xsAPI.c              |   45 +-
 xs/sources/xsAll.h              |   20 +-
 xs/sources/xsArray.c            |   17 +-
 xs/sources/xsBigInt.c           |    6 +-
 xs/sources/xsCode.c             |  155 +-
 xs/sources/xsCommon.c           |  378 +--
 xs/sources/xsCommon.h           |   35 +-
 xs/sources/xsDataView.c         |   25 +-
 xs/sources/xsDate.c             |    7 +-
 xs/sources/xsDebug.c            |  171 +-
 xs/sources/xsFunction.c         |   45 +-
 xs/sources/xsGenerator.c        |   53 +-
 xs/sources/xsGlobal.c           |   44 +-
 xs/sources/xsJSON.c             |  118 +-
 xs/sources/xsLexical.c          |  130 +-
 xs/sources/xsMath.c             |    8 +-
 xs/sources/xsMemory.c           |  192 +-
 xs/sources/xsModule.c           | 1189 ++++---
 xs/sources/xsNumber.c           |    2 +-
 xs/sources/xsObject.c           |   38 +-
 xs/sources/xsPlatforms.c        |  264 +-
 xs/sources/xsPromise.c          |   22 +-
 xs/sources/xsProperty.c         |   41 +-
 xs/sources/xsRegExp.c           |  144 +-
 xs/sources/xsRun.c              |   43 +-
 xs/sources/xsScope.c            |   12 +-
 xs/sources/xsScript.c           |   30 +
 xs/sources/xsScript.h           |    7 +-
 xs/sources/xsSnapshot.c         |   26 +-
 xs/sources/xsSourceMap.c        |    3 +
 xs/sources/xsString.c           | 1719 ++++++++--
 xs/sources/xsSyntaxical.c       |   36 +-
 xs/sources/xsTree.c             |    4 +-
 xs/sources/xsType.c             |   35 +-
 xs/sources/xsre.c               | 6579 +++++++++++++++++++++------------------
 xs/tools/unicode/cases.js       |  261 ++
 xs/tools/unicode/download.sh    |   17 +
 xs/tools/unicode/forms.js       |  469 +++
 xs/tools/unicode/identifiers.js |   82 +
 xs/tools/unicode/manifest.json  |   26 +
 xs/tools/unicode/properties.js  |  115 +
 xs/tools/unicode/tests.js       |   31 +
 xs/tools/xsc.c                  |   90 +-
 xs/tools/xsl.c                  |  121 +-
 xs/tools/xslBase.c              |    7 +-
 xs/tools/xslSlot.c              |   56 +-
 xs/tools/xslStrip.c             |    3 +
 xs/tools/xst.c                  | 1133 +++++--
 xs/tools/xst.h                  |   12 +-
 67 files changed, 11917 insertions(+), 5758 deletions(-)
```

### git log

`git log --oneline 55cc73f..d7895e4 xs/`

```
82faa614 enable FUZZILLI based on environment variable
b1a957a5 XS snapshot: restore chunks in case of exceptions
5d07a6b9 Merge branch 'master' into pico
aed12c0c comment with make flags for fuzzilli
7b22c0b8 integrate Fuzzilli REPRL. Based on original work by @jessysaurusrex #817
45b57f83 Merge branch 'master' into pico
16b004cf XS: sort degenerate cases
9f82ed5f Merge branch 'master' into pico
25973d7a eliminate cross-task interference on instrumentation update
8bd7a2fd Timer.clear silently ignores undefined and null (matches clearTimeout)
90ee6066 XS: report unhandled rejection reason
cbddd769 add mxUnalignedAccess for esp8266
d249c2c9 Merge branch 'master' into pico
9c134d8e XS: freeze Compartment Function eval after lockdown
211daa54 update to latest
25356114 XS: fix Compartment Function constructors
fe6fc8ae Merge branch 'master' into pico
e2062b07 XS compartment modules: meta vs redirect
715d45dc esp32s3 fix for idf4.4
12cda75b XS linker: #841
214d0630 XS:  import.meta and loadNowKook
c5c7744d take 2
937e710a MODDEF_XS_DONT_INITIALIZE_TIME (h/t @cmidgley)
b3c4391e XS: compartment resolve hook and import.meta
76a2bd9a stop on first error to avoid crashing in unexpected parser state later #826
4f966188 XS: #818
df73b641 xst: use smaller stack for threads
bba8d931 XS: deep freezing does not throw
478faeca XS: regression tests
37b3be17 Merge branch 'master' into XS11.7
f0991812 XS: #815
e140f296 XS: #807
cf213e82 XS: #805
5129440b XS: #797
5dc758ea XS: #797
057413b3 XS:  mark closures non recursively
6b0734ce XS: #798
dcb4456c XS: #789
6e15c403 XS: #788
7301b791 XS: lower/upper unicode table size
ab28c902 xst: Linux and Windows builds
b77b38a4 xst: TextDecoder/Encoder
ecf9e4c8 add HostBuffer to ArrayBufferTypes, define BufferLike for Moddable SDK APIs, start to use BufferLike (with help from @cmidgley)
d130ef57 XS: normalize strings
e69044c7 XS: build xsc/xsl
9851a261 XS: normalize tests
3c2a82ba XS: normalize strings
eadbf1cc XS: normalize strings
09e9b09e XS: regexp unicode properties
3bfbd870 Merge branch 'master' into XS11.7
18cdd618 XS: Compartment constructor in preparation
3a7393bf XS: jessy 1
9d418865 XS: jessy 11
eebf0ab6 XS: jessy 16
ed7c204b XS: jessy 10
cbae5ae2 XS: jessy 5
fa308e6d XS: jessy 3
8b13f7d7 XS: jessy 2
fc280384 XS: #768
3bfc64a7 XS: do not source map twice!
25b8ccd3 XS: ?.keyword
03ddfa39 XS: fxUTF8Compare
1708e71a XS: zero character, xsbug
091afdc9 XS: zero character, fxUTF8Compare
33226962 XS: unicode tool
c999049d XS: unicode tool
7ff9176b XS: toLowerCase and toUpperCase
82eac146 XS: RegExp test262
b5274043 XS: toLowerCase and toUpperCase
d70acdc4 XS: toLowerCase and toUpperCase
090eab76 if source file referenced by source map is unavailable, fallback to original source file for debugging path/lines #771
cc9c675a xsbug, mcsim, test262 app on Windows
ffac1d53 Merge branch 'master' into XS11.7
9cf05690 XS: #768
6ebd4e89 warn on missing source map (instead of fail) #771
8e94a366 XS: #769 (from Patrick)
d14207ba include mc.defines.h so MODDEF_* is available to XS sources
cdf1c9aa Merge branch 'master' into XS11.7
7bd3d01c XS: CESU-8 backup
c40aa0dd XS: CESU-8 backup
971d3bd2 XS: CESU-8 backup
0c4403b0 XS: #774
9a468e89 XS: CESU-8 backup
7b48fc15 XS: CESU-8 backup
4728631c XS: CESU-8 backup
8074d4ea XS: MODDEF_XS_TEST
ad4745d8 migrate shared code from xsHost to xsHosts
dc9e9364 XS: module descriptor
80201616 XS: Compartment tests
92b45711 XS linker accessor strip
43cc0296 XS: Compartment tests
5b6f857d XS: Compartment tests
3b7cfd24 XS: Compartment globalLexicals
3ef6c22b for xsnap
5f8ba4f8 XS: module map in xsbug
ec38628f XS: linux and windows build
cccd8e42 XS: apps and mods specifiers
ce937b4e XS: #760
f000fe13 XS: #759
4a1e972c XS: #748
bd3a1861 Merge branch 'master' into XS11.6
24c94d97 xst results and conformance document
40fb59ae xst timers and jobs
127c715d XS: RegExp 'd' flag
df45e15e XS: ignore iterator return exception in catch
f30e8fdd XS: Date clips -0
dfe2e6ee XS: missing unscopables
adf9e4e7 XS: DataView constructor with resizable buffer throws RangeError instead of TypeError
dbaf2b90 XS: partial descriptor in Object.freeze and Object.seal
a6d89fb8 XS: legacy decimal integer
b4f33472 XS: no caller property in method in object literal
fa3182f0 XS: primitive __proto__ in object literal
80397cf8 XS: arguments in eval in arrow function in field initializer
bebabc9e XS: finally completion value
7bf571cc XS: legacy octal escape
eaf479f5 XS: identifiers unicode 14
d48bcdd4 xst: test262 multiple threads
99df7f0e XS: function name
f597570f XS: function name
c782c769 Merge branch 'master' into XS11.6
9aa61c8e XS: static module record bindings
dc02cb95 XS shared modules
0b0046bb XS modules backup
e1bd4b52 XS: mxNoChunks sweep, mxPoisonSlots
57cb2b87 Merge branch 'master' into XS11.6
df98766f Merge branch 'master' into XS11.6
cfda237c XS static module record from archive
c1a48f14 Merge branch 'master' into XS11.6
efed4359 XS: remove bytecode option from StaticModuleRecord
61b058ba XS: module specifier resolution
024322b6 XS compartment module map and preparation
52e7f8a7 XS: alien static module record
a86b9a49 XS: alien static module record
e4509820 XS: compartment global lexical variables, alien static module record
2d53a57a Merge branch 'master' into pico
82b70c25 Merge branch 'master' into pico
f999e6e8 Merge branch 'master' into pico
518c9fb5 Merge branch 'master' into pico
3043d999 chunk allocator should return NULL on failure rather than throw
3ca9b114 Merge branch 'master' into pico
6ba90068 Merge branch 'master' into pico
33083479 Merge remote-tracking branch 'origin/master' into pico
bf2c0fb4 mods
68148260 workers like esp
f8f91b2e Merge remote-tracking branch 'origin/master' into pico
9fc965b2 Merge remote-tracking branch 'origin/master' into pico
43bcf83b Merge branch 'master' into pico
f408c4f2 XS: #592
9aefa610 XS: marshalling
79004df0 XS linker explicit strip REPL fix
4345a257 XS linker explicit strip REPL fix
6c17c814 XS: error stack at creation - REBUILD ALL FROM SCRATCH
7a745029 put back workaround for best_effort_wfe_or_timeout hang
cbb36337 remove workaround for best_effort_wfe_or_timeout hang
01b2d2c9 merge in eval debugging
46bee599 ili9341 with xpt2046, add protection around flash calls
9efd82ef preferences. ft6206 touch.
d71782b7 Merge remote-tracking branch 'origin/master' into pico
9d9eced0 Merge remote-tracking branch 'origin/master' into pico
94052de4 spi and ili9341 on pico
2c373c31 can simplify use of best_effort_wfe_or_timeout
a08a7d52 use tud_cdc_rx_cb to process serial data for xsbug sooner
10737fbb get modMessagePostToMachineFromISR working
e042d2cb give linux more time to connect to xsbug
2d20e35e define mxUnalignedAccess in xsPlatform.h to remove platform dependency in xsRun.c
4a1b9eab runloop sleep
45b91ffb use USB for xsbug connection
de9e2382 helloworld works on pico blink and button work with change to pin number in the example
6f1bb910 Merge remote-tracking branch 'origin/master' into pico
0d56425a first pico
```

**Review by dckc — CHANGES_REQUESTED** [review]
I think `METER_TYPE` needs to be bumped.

And I wonder if the mac OS minimum version bump carries any documentation burden.

Otherwise, all looks in order.

File: packages/xsnap/test/test-xs-perf.js (hunk 1 above)
Comment (kriskowal) [reply packages/xsnap/test/test-xs-perf.js:52]: Posting a commit.

**Comment by kriskowal** [comment]
Per https://github.com/Agoric/agoric-sdk/pull/4832#issuecomment-1075797199, I sniffed around in XS and found `mxRun` gets defined in many `make` targets, including xsnap’s and xsnap-worker’s. My intuition is that it distinguishes programs with a main loop, without which some object code can be omitted. I found no comments hinting at its purpose. I defer to @phoddie.

I found no indication that we document our Mac OS compatibility range yet. We probably ought to add it somewhere. If so,  the minimum was Lion (circa 2011) and is now Yosemite (circa 2017).

Otherwise, I bumped the `METER_TYPE` and took care to make sure it occurs in an isolated breaking change commit.

**Comment by phoddie** [comment]
`mxRun` is almost always set to 1. The exceptions are tools like `xsc` (the XS compiler) which use XS in unusual ways (xsc doesn't run scripts but compiles them to byte-code). You absolutely want to set `mxRun` to 1 for xsnap.

> I found no indication that we document our Mac OS compatibility range yet.

As a rule for the Moddable SDK on macOS, we support the current version and one before it. Mac users generally keep their machines pretty up-to-date, so that isn't a problem. On the other hand, macOS compatibility is quite solid so we don't aggressively turn off older versions.

**Review by dckc — APPROVED** [review]
:shipit:

**Review by warner — APPROVED** [review]
sounds good

**Pull request merged**